RFC006 - Release Planning & Management Process (#12)
thegooddocsproject/request-for-comment: rfc006-release-planning-mgt.md

1. The PSC will evaluate each Working Group's release plan and provide feedback. As the plan will be based on a realistic bandwidth estimate, the PSC should focus on identifying items to be added/removed based on strategic priorities. The PSC and Working Groups will negotiate and come to a consensus on what from each sub-project will be included in the release.
1. The Release Manager will track progress towards the release's goals as reported in ZenHub through regular iterations (sprints), and will be responsible for identifying slippage and risk. The Release Manager will provide regular updates to the project's members as part of the regular meetings, although overall progress will be available anytime from ZenHub for members with access.
1. If the release appears at risk for slippage, the Release Manager will coordinate with the Working Groups to see if progress can be brought back to the baseline, or if items will need to be removed from the release. The Release Manager will inform the PSC if the latter occurs, to determine if this is feasible, and if not how it should be handled.

> 1. Working Groups will provide an update/demo of the results of their work in preparation for the release. Once the PSC and other stakeholders are satisfied that it meeds the commitment, members of the project can perform the necessary reviews/approvals, merge pull requests, and package the Working Groups' deliverables.

Reviewer: * Typo: s/meeds/needs/
* I suggest setting up a heading for "Roles and Responsibilities", which includes "Release manager" (as you have been articulating here), and also "Section contributor" (or similar better title), which might be a group rather than a person.
* We should then spell out the responsibilities of this coordinator, which would include "providing updates which align with Good Docs standards", and providing changelog updates for the Release Manager to collate.

Author: Fixed typo, and added a Roles & Responsibilities section.

- GitHub: Work towards the goals of the Release will be represented by Issues in GitHub. GitHub will also be used to distribute the end products, either directly from source repositories, or as packages (e.g. .ZIP files) for end users to download from the **Releases** page of the appropriate repository.
- ZenHub: ZenHub will provide organization and visibility through its ability to group individual work items (e.g. Issues) into higher-level Epics, and show progress toward their completion.

## In Scope

All Working Groups and sub-projects of The Good Docs Project are eligible for inclusion in a Release. That said, the precise contents of any Release should be reached through consensus between the PSC, who should seek to maximize the value delivered, and the Working Groups, who should ensure commitments are feasible.

## Out of Scope

The following items are out of scope for this proposal, as they're either not integral to a Release Planning & Management plan, or are addressed by other (potentially upcoming) RFCs:

> - Versioning: RFC-008 will detail the versioning or numbering standard for the project.

Reviewer: I feel that our versioning strategy should be a prerequisite for this step, and we should be referencing it from this RFC.

Author: Added RFC-008 (Versioning) as a dependency for this one, although in my mind they're co-dependencies. It's something of a chicken and an egg problem, how to consider releases until you know how you'll version them, but how do you apply versions when you don't know what you're releasing?

# RFC006 - Release Planning & Management Process

## Proposed by

Aaron Peters (@acpkendo)

## Current status

- [x] Draft
- [x] Under discussion (until 2021-09-29)
- [ ] Final comment and voting (until YYYY-MM-DD)
- [ ] Accepted
- [ ] Rejected
- [ ] Implemented
- [ ] Deferred
- [ ] Withdrawn

## Proposal overview

This proposal outlines a process to plan for and manage releases for The Good Docs Project. Specifically, it will detail:

- The process through which the project defines and agrees on a Release;
- The versioning scheme the project will use to track and communicate progress across Releases, and;
- The tooling the project will use to manage progress towards Releases.

## Motivation for Releases & a Release Planning Strategy

Many volunteer open source projects will release new versions of their output (software) whenever it's deemed to be in a user-ready state that includes new value. But The Good Docs Project has identified a few prevailing reasons for creating formal Releases of our outputs, as follows:

- For technical writers, creating formal releases makes it easy to ensure they're always working from the latest stable/approved versions of templates.

> - For integrators (i.e. those who might create solutions based on our work), well-versioned releases make it easy to ensure (backward-)compatibility.

Reviewer: ```suggestion
- For integrators (i.e. those who might create solutions based on our work), well-versioned releases make it easy to ensure (backward-)compatibility and plan future development of their downstream products.
```

## Definition of a Release of The Good Docs Project

As an organization with multiple sub-projects (some of which arguably have their own sub-projects) working with different cadences and outputs, we'll define for the purposes of the process this RFC describes what constitutes a "release" from The Good Docs Project. It's been noted that The Good Docs Project will produce two types of outputs:

- "Product Oututs," or assets such as templates that will actually be used by our core audience in their own documentation efforts, and;
- "Non-Product Outputs," which includes thought leadership, whitepapers, and other content that may inform documentation efforts but will not be used directly.

> After discussion among the PSC and community, we're defining a Release as **a snapshot of Product Outputs as agreed upon by the PSC and Working Groups**. At the time of writing, the following outputs are considered Product Outputs of TGDP:

Reviewer: ```suggestion
A release constitutes:
* A list of products and features which are to be included in the release.
* A definition of quality to be applied to the release, such as "stable", "beta", "bleeding edge", as defined by our internal processes. All proposed features will be expected to meet this criteria, otherwise it won't be included.
* A statement of planned support that will be applied to the release (which might be "no support", or "security and typo patches").
* A release date, which will typically align with a key event, such as a conference.
* A working schedule with key milestones. A sub-project will be expected to meet the milestones or risk having proposed updates being excluded from the release.

At the time of writing, the following outputs are considered Product Outputs of TGDP:
```

Reviewer: You can refer to the [OSGeoLive schedule](https://docs.google.com/spreadsheets/d/1kO6zzmLFfprZGgp5x7Sjwi-EVN6NTGDR4KXvFVtNpR0/edit?hl=en_GB&hl=en_GB&pli=1#gid=0) which is a spreadsheet of key milestones for each of our releases. These dates were not fixed in stone, we would often adjust the dates as we worked through the schedule. Then for future releases we would copy the prior schedule, making adjustments from prior learnings, and known upcoming holidays or scope changes.

I gave a presentation covering our OSGeoLive build process at: http://cameronshorter.blogspot.com/2011/06/memoirs-of-cat-herder-coordinating.html

- The Templates collection
- The Doc Tools initiatives
- The Chronologue project, by virtue of its inclusion as examples to both the above

Note however that while a given Release from The Good Docs Project may not include any *updated* material as part of a Product Output, its current version will still be included as part of any release packages that are creation. For example, the Doc Tools Working Group may release a new version of that software as part of a Release implementing a new feature. But the Chronologue WG's content included with that software need not be updated at all, unless the new feature specifically requires it (e.g. it makes use of a new metadata field).

## Roles and Responsibilities

The main participants in the Release Planning Process are as follows:

> - Project Steering Committee: The PSC will collectively create goals for releases, and 

Reviewer: ```suggestion
- Project Steering Committee: The PSC will review and approve proposed releases. 
```

I suggest that the role of the PSC is oversight and review. Sure, PSC members can suggest goals and dates to the release manager, but I suggest that it is the release manager's responsibility to drive this.

The work of selecting a date, setting a schedule, selecting a quality standard, is something that can be set up best by one person rather than a committee.

Author: I feel like reducing the PSC to purely oversight will exclude PSC members who aren't also part of a WG that creates Product Output from weighing in on what they'd want to see. But I re-worded to accommodate the spirit of the above I think.

> - Release Manager: This role will be responsible for driving the planning process for future Releases, and for maintaining an easy-to-understand status on any in-process Release(s). This individual will interact regularly with Working Group leaders to understand current progress, and update the PSC if it's trending away from targets.

Reviewer: ```suggestion
- Release Manager: This role will be responsible for:
  - Driving the planning process for future releases;
  - Selecting a release version number and associated quality bar for the release;
  - Proposing a release date, schedule milestones;
  - Negotiating scope with Working Group leaders and PSC; 
  - Maintaining an easy-to-understand status on any in-process Release(s);
  - Interacting regularly with Working Group leaders to understand current progress, and update the PSC if it's trending away from targets.
```

## Proposal

The release planning process will be a collaborative process between members, their Working Groups, and the Project Steering Committee, and the Release Manager as follows:

1. The PSC will decide on a target date for an upcoming release. This can be driven by the desire to release together with a particular event (e.g. at a Write The Docs conference), or based on a regular schedule (such as quarterly). The overall goals of the project release will be tracked in a TBD location: ZenHub if it supports recursive Epics, or e.g. a wiki if not.
1. Based on a target release date, Working Groups will convene and decide (based on capacity) the items to which they can commit. They will be able to plan the precise issues they will complete as part of a release using their workspace in the ZenHub application.

> 1. The PSC will evaluate each Working Group's release plan and provide feedback. As the plan will be based on a realistic bandwidth estimate, the PSC should focus on identifying items to be added/removed based on strategic priorities. The PSC and Working Groups will negotiate and come to a consensus on what from each sub-project will be included in the release.

Reviewer: As above, I feel the PSC's role should only be oversight, only needing to review, comment and then approve proposals put to them.

Day to day work, such as reviewing a working group's proposal should be done by the release manager (who may defer to the PSC for a contentious decision if needed).

If the release manager job becomes too big for one person, we could spawn a Release Management Committee in order to spread the load and decision making.

(This comment also applies to other lines in this Proposal section.)

Author: As above, I think wording this to explicitly limit the PSC and its members to oversight is too strict. PSC members, in the interest of steering the project, should be a part of the overall goal-setting process. But making the PSC's oversight role more explicit is a good idea.

## Toolset

In as described above, this process will leverage three major tools to plan and manage releases.

- Wiki: The high-level goals and commitments of a Release should be documented on a publicly-available wiki, to align expectations and provide transparency. The Release Manager will collect proposals from the Working Groups, and facilitate the discussions between Working Groups and the PSC to reach a consensus on the Release's contents.
- GitHub: Work towards the goals of the Release will be represented by Issues in GitHub. GitHub will also be used to distribute the end products, either directly from source repositories, or as packages (e.g. .ZIP files) for end users to download from the **Releases** page of the appropriate repository.
- ZenHub: ZenHub will provide organization and visibility through its ability to group individual work items (e.g. Issues) into higher-level Epics, and show progress toward their completion.

## In Scope

> All Working Groups and sub-projects of The Good Docs Project are eligible for inclusion in a Release. That said, the precise contents of any Release should be reached through consensus between the PSC, who should seek to maximize the value delivered, and the Working Groups, who should ensure commitments are feasible.

Reviewer: ```suggestion
All Working Groups and sub-projects of The Good Docs Project are eligible for inclusion in a Release. That said, the precise contents of any Release should be reached through consensus between the PSC, who should seek to maximize the value delivered, and the Working Groups, who should ensure commitments are feasible and our quality standards are maintained.
```

## Dependencies

This RFC should be considered together with RFC-008, which describes in detail the versioning and numering scheme to be used for the project's releases.

## Out of Scope

The following items are out of scope for this proposal, as they're either not integral to a Release Planning & Management plan, or are addressed by other (potentially upcoming) RFCs:

> - Promotion/Marketing: This is a more holistic topic of who should drive promotion and marketing for the project. While Releases should certainly be part of this, the individual who takes on this role should have 

Reviewer: Want to finish this sentence?

- For project members, it aligns expectations on when to target tasks for completion, and way to ensure only updates of the appropriate value and quality are distributed to users.

The Good Docs Project has grown to contain multiple sub-initiatives ranging from the templates (arguably the core of the project's activities at time of writing), to tooling aimed at technical writers, to useful content in the form of example documentation and blog posts. Given the project's increasing membership across multiple time zones, implementing a standard process through which the Project Steering Committee and Working Groups can define and manage releases will align expectations and better enable monitoring of progress.

## Definition of a Release of The Good Docs Project

As an organization with multiple sub-projects (some of which arguably have their own sub-projects) working with different cadences and outputs, we'll define for the purposes of the process this RFC describes what constitutes a "release" from The Good Docs Project. It's been noted that The Good Docs Project will produce two types of outputs:

> - "Product Outputs," or assets such as templates that will actually be used by our core audience in their own documentation efforts, and;

Reviewer 2: "Directly" not "actually".

> - "Non-Product Outputs," which includes thought leadership, whitepapers, and other content that may inform documentation efforts but will not be used directly.

Reviewer 2: Explicitly include blog posts and outreach efforts, please

# RFC006 - Release Planning & Management Process

## Proposed by

Aaron Peters (@acpkendo)

## Current status

- [x] Draft
- [x] Under discussion (until 2021-09-15)
- [ ] Final comment and voting (until YYYY-MM-DD) {{Add date after selecting this status.}}
- [ ] Accepted
- [ ] Rejected
- [ ] Implemented
- [ ] Deferred
- [ ] Withdrawn

## Proposal overview

This proposal outlines a process to plan for and manage releases for The Good Docs Project. Specifically, it will detail:

- The process through which the project defines and agrees on a Release;
- The versioning scheme the project will use to track and communicate progress across Releases, and;

> - The tooling the project will use to manage progress towards Releases.

Reviewer 3: 💭 ❔ __thoughts/questions:_ Thought I'd kick us off 💨 

- I have questions around many areas I feel should be included so like @barbaricyawps did in her #10 she noted what was going to be out of scope which helped curtail some of those questions I immediately had.
- ❔ Which projects will need to be included in the Release Coordination? How do those Projects/Working-groups communicate that they need to be added to the Release process for any upcoming releases.  Or can a working-group release on their own cadence independently?  Are there rules around how all this works/communicated?
- ➡️ Versioning Scheme -- I didn't see any concrete notes on the process here in the proposal so not sure what is in scope/out-of-scope.  But think this should potentially be included here.  Happy to provide language/thoughts here as this is probably in my wheel-house.
- ➡️ Tooling, other than the mention of Zenhub as our project-management.  A simple version of this might be bundling/tagging releases on github.  And that our default tooling is github's packaging.
- 📣 How do we communicate releases to external/consumers and the _marketing-team_.  If we are setting a regular cadence of releases we should communicate that.
- 💭 Btw, I'm a fan of [keep-a-changelog](https://keepachangelog.com/en/1.0.0/) as one way to manage some release-notes even if there are deliverables.

Author: Hi @mgan59, I made some changes to facilitate discussion, a fair amount of which I feel will still be required.

First question: I think all WGs/sub-projects should be eligible to be included in a release, but I think it's a mistake to say "all of them must" or "releases consist of precisely X." An Agile approach to this would allow the individual teams to decide what they think is the most valuable items (within their own purview) are, as well as what they can confidently commit to in order to meet a given release (most likely time-bounded, to speak to the point raised by @barbaricyawps below). But making requirements such as "all releases contain updates to the templates" or "every WG must contribute something to a release" will inevitably put us in a position where e.g. a WG is short on help, can't get a quality contribution ready in time, and then what? IMO each release should be a negotiated agreement between the PSC (i.e. "the business") on what they want to reach the project's high-level goals, and the WGs (i.e. the Agile teams) on what they feel is most valuable and can deliver.

Second point: Versioning... I've volunteered to draft a separate RFC for this, which in my mind makes sense as the number attached to deliverables isn't dependent on the process used to plan for those deliverables, and vice versa. Once I'm ready to start on this draft I will certainly work with you on its details.

Third point: Tooling... I added a section to describe the toolset and its roles more clear, although it probably needs a little more yet.

Fourth point: Marketing... This should arguably be a different RFC, as it's a different function performed by a different person (Product Marketing Manager) in many organizations. I also feel the organization would benefit from some sort of marketing/PR chair/team to promote the project across a variety of channels. Stated more simply, I feel like this is something the marketing team (once it exists) should drive.
